design/handle-backup-of-volumes-by-resources-filters.md

```diff
@@ -0,0 +1,246 @@
+
+#### Special rule definitions:
+- A attribute selector in `AttributeSelectors` with a specific key and empty value, which means the value matches any value. For example, if the `conditions.nfs` is `{}`, it means if `NFS` is used as `persistentVolumeSource` in Persistent Volume will be skipped no matter what the NFS server or NFS Path is.
+ 
+- The size of each single filter value should limit to 256 bytes in case of an unfriendly long variable assignment.
+ 
+- For capacity for PV or size for Volume, the value should include the lower value and upper value concatenated by commas. And it has several combinations below:
+  - "0,5Gi" or "0Gi,5Gi" which means capacity or size matches from 0 to 5Gi, including value 0 and value 5Gi
+  - ",5Gi" which is equal to "0,5Gi"
+  - "5Gi," which means capacity or size matches larger than 5Gi, including value 5Gi
+  - "5Gi" which is not supported and will be failed in validating configuration.
+ 
+### Configmap Reference
+Currently, resources policies are defined in `BackupSpec` struct, it will be more and more bloated with adding more and more filters which makes the size of `Backup` CR bigger and bigger, so we want to store the resources policies in configmap, and `Backup` CRD reference to current configmap.
+
+the `configmap` would be like this:
+```yaml
+apiVersion: v1
+data:
+ policies.yaml:
+  ----
+  version: v1
+  volumePolicies:
+  - conditions:
+      capacity: "0,100Gi"
+      csi:
+        driver: aws.ebs.csi.driver
+        fsType: ext4
+      storageClass:
+      - gp2
+      - ebs-sc
+    action:
+      type: volume-snapshot
+      parameters:
+        volume-snapshot-timeout: "6h"
+kind: ConfigMap
+metadata:
+ creationTimestamp: "2023-01-16T14:08:12Z"
+ name: backup01
+ namespace: velero
+ resourceVersion: "17891025"
+ uid: b73e7f76-fc9e-4e72-8e2e-79db717fe9f1
+```
+
+A new variable `resourcePolices` would be added into `BackupSpec`, it's value is assigned with the current resources policy configmap
+```yaml
+apiVersion: velero.io/v1
+kind: Backup
+metadata:
+  name: backup-1
+  spec:
+    resourcePolices:
+      refType: Configmap
+      ref: backup01
+    ...
+```
+
+The configmap basically equivalent to what is generated by the command `kubectl create cm backup01 --from-file policies.yaml`
+ 
+The configmap only stores those assigned values, not the whole resources polices.
+ 
+The name of the configmap is `$BackupName`, and it's in Velero install namespace.
+ 
+#### Life-cycle of resources policies configmap
+- Resource policies configmap only been generated once Velero backup command with flag `backup-policy-configmap` to import configuration.
+- Resource policies configmap could be referenced by another backup in Velero backup command with the flag `policy-from-backup` to reference configuration in another backup.
+- If all referenced backup has been deleted, the relevant configmap should be removed.
```

Comment:
Let's only delete the configmap if it was created by velero using the `backup-policies-from-file` since in other cases the lifecycle may be managed by a third party operator.

Reply:
I've updated my design now.

The resource policies configmap is maintained by users. which would be more flexible. users could create or delete configmap through `kubectl`

Velero will provide one flag `reousce-pilicies-configmap` to refer to resource policies configmap created by users when using the command `velero backup create`